File: frontend/flutter/lib/core/widgets/app_button.dart
import 'package:flutter/material.dart';
import '../theme/design_tokens.dart';

enum ButtonVariant { primary, secondary, tertiary, destructive, ghost }

enum ButtonSize { small, medium, large }

class AppButton extends StatefulWidget {
  final String text;
  final VoidCallback? onPressed;
  final ButtonVariant variant;
  final bool isLoading;
  final bool isFullWidth;
  final IconData? icon;
  final double? width;
  final double? height;
  final ButtonSize size;

  const AppButton({
    super.key,
    required this.text,
    this.onPressed,
    this.variant = ButtonVariant.primary,
    this.isLoading = false,
    this.isFullWidth = false,
    this.icon,
    this.width,
    this.height,
    this.size = ButtonSize.medium,
  });

  @override
  State<AppButton> createState() => _AppButtonState();
}

class _AppButtonState extends State<AppButton> {
  bool _isPressed = false;

  @override
  Widget build(BuildContext context) {
    final ButtonStyle style = _getButtonStyle();
    final double buttonHeight = _getButtonHeight();
    final Widget child = widget.isLoading
        ? SizedBox(
            height: _getIconSize(),
            width: _getIconSize(),
            child: const CircularProgressIndicator(
              strokeWidth: 2,
              valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
            ),
          )
        : _buildContent();

    return AnimatedScale(
      scale: _isPressed ? 0.95 : 1.0,
      duration: const Duration(milliseconds: 100),
      child: GestureDetector(
        onTapDown: (_) => setState(() => _isPressed = true),
        onTapUp: (_) => setState(() => _isPressed = false),
        onTapCancel: () => setState(() => _isPressed = false),
        child: SizedBox(
          width: widget.isFullWidth ? double.infinity : widget.width,
          height: widget.height ?? buttonHeight,
          child: _buildButton(style, child),
        ),
      ),
    );
  }

  Widget _buildButton(ButtonStyle style, Widget child) {
    // Use different button types based on variant
    if (widget.variant == ButtonVariant.tertiary) {
      return widget.icon != null && !widget.isLoading
          ? OutlinedButton.icon(
              onPressed: widget.onPressed,
              icon: Icon(widget.icon, size: _getIconSize()),
              label: Text(widget.text),
              style: style,
            )
          : OutlinedButton(
              onPressed: widget.isLoading ? null : widget.onPressed,
              style: style,
              child: child,
            );
    } else if (widget.variant == ButtonVariant.ghost) {
      return widget.icon != null && !widget.isLoading
          ? TextButton.icon(
              onPressed: widget.onPressed,
              icon: Icon(widget.icon, size: _getIconSize()),
              label: Text(widget.text),
              style: style,
            )
          : TextButton(
              onPressed: widget.isLoading ? null : widget.onPressed,
              style: style,
              child: child,
            );
    } else {
      return widget.icon != null && !widget.isLoading
          ? ElevatedButton.icon(
              onPressed: widget.onPressed,
              icon: Icon(widget.icon, size: _getIconSize()),
              label: Text(widget.text),
              style: style,
            )
          : ElevatedButton(
              onPressed: widget.isLoading ? null : widget.onPressed,
              style: style,
              child: child,
            );
    }
  }

  Widget _buildContent() {
    return Text(
      widget.text,
      style: TextStyle(fontSize: _getFontSize(), fontWeight: FontWeight.w600),
    );
  }

  double _getButtonHeight() {
    switch (widget.size) {
      case ButtonSize.small:
        return 48.0; // Min touch target 48dp (WCAG AA)
      case ButtonSize.medium:
        return 48.0;
      case ButtonSize.large:
        return 56.0;
    }
  }

  double _getFontSize() {
    switch (widget.size) {
      case ButtonSize.small:
        return 14.0;
      case ButtonSize.medium:
        return 16.0;
      case ButtonSize.large:
        return 18.0;
    }
  }

  double _getIconSize() {
    switch (widget.size) {
      case ButtonSize.small:
        return 16.0;
      case ButtonSize.medium:
        return 20.0;
      case ButtonSize.large:
        return 24.0;
    }
  }

  ButtonStyle _getButtonStyle() {
    final theme = Theme.of(context);
    final colorScheme = theme.colorScheme;

    switch (widget.variant) {
      case ButtonVariant.primary:
        return ElevatedButton.styleFrom(
          backgroundColor: colorScheme.primary,
          foregroundColor: colorScheme.onPrimary,
          disabledBackgroundColor: colorScheme.primary.withValues(alpha: 0.38),
          disabledForegroundColor: colorScheme.onPrimary.withValues(
            alpha: 0.38,
          ),
          padding: _getButtonPadding(),
          elevation: 0, // WhatsApp flat design
          shape: RoundedRectangleBorder(
            borderRadius: AppRadius.button,
          ),
        );
      case ButtonVariant.secondary:
        return ElevatedButton.styleFrom(
          backgroundColor: colorScheme.secondary,
          foregroundColor: colorScheme.onSecondary,
          disabledBackgroundColor: colorScheme.secondary.withValues(
            alpha: 0.38,
          ),
          disabledForegroundColor: colorScheme.onSecondary.withValues(
            alpha: 0.38,
          ),
          padding: _getButtonPadding(),
          elevation: 0, // WhatsApp flat design
          shape: RoundedRectangleBorder(
            borderRadius: AppRadius.button,
          ),
        );
      case ButtonVariant.tertiary:
        return OutlinedButton.styleFrom(
          backgroundColor: Colors.transparent,
          foregroundColor: colorScheme.primary,
          disabledForegroundColor: colorScheme.onSurface.withValues(
            alpha: 0.38,
          ),
          side: BorderSide(
            color: widget.onPressed != null && !widget.isLoading
                ? colorScheme.outline
                : colorScheme.outline.withValues(alpha: 0.38),
            width: 1.0,
          ),
          padding: _getButtonPadding(),
          elevation: 0, // WhatsApp flat design
          shape: RoundedRectangleBorder(
            borderRadius: AppRadius.button,
          ),
        );
      case ButtonVariant.destructive:
        return ElevatedButton.styleFrom(
          backgroundColor: colorScheme.error,
          foregroundColor: colorScheme.onError,
          disabledBackgroundColor: colorScheme.error.withValues(alpha: 0.38),
          disabledForegroundColor: colorScheme.onError.withValues(alpha: 0.38),
          padding: _getButtonPadding(),
          elevation: 0, // WhatsApp flat design
          shape: RoundedRectangleBorder(
            borderRadius: AppRadius.button,
          ),
        );
      case ButtonVariant.ghost:
        return TextButton.styleFrom(
          backgroundColor: Colors.transparent,
          foregroundColor: colorScheme.primary,
          disabledForegroundColor: colorScheme.onSurface.withValues(
            alpha: 0.38,
          ),
          padding: _getButtonPadding(),
          shape: RoundedRectangleBorder(
            borderRadius: AppRadius.button,
          ),
        );
    }
  }

  EdgeInsets _getButtonPadding() {
    switch (widget.size) {
      case ButtonSize.small:
        return const EdgeInsets.symmetric(horizontal: 12, vertical: 8);
      case ButtonSize.medium:
        return const EdgeInsets.symmetric(horizontal: 16, vertical: 12);
      case ButtonSize.large:
        return const EdgeInsets.symmetric(horizontal: 24, vertical: 16);
    }
  }
}

File: frontend/android-native/app/src/main/java/com/example/minitasktracker/core/ui/components/AppButton.kt
package com.example.minitasktracker.core.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.example.minitasktracker.core.ui.theme.*

/**
 * TekTech AppButton Component
 * 
 * Unified button component with 5 variants and all states
 * - Primary: Main actions (Indigo 600)
 * - Secondary: Secondary actions (Green 500)
 * - Tertiary: Less prominent actions (outlined)
 * - Destructive: Dangerous actions (Red 500)
 * - Ghost: Minimal actions (text only)
 * 
 * States: Default, Pressed, Disabled, Loading
 * Accessibility: 48dp minimum touch target, semantic labels
 * 
 * @param text Button label
 * @param onClick Click handler
 * @param modifier Modifier
 * @param variant Button style variant
 * @param size Button size (small, medium, large)
 * @param enabled Whether button is enabled
 * @param loading Whether button shows loading spinner
 * @param icon Optional leading icon
 * @param fullWidth Whether button fills max width
 */

enum class ButtonVariant {
  PRIMARY,
  SECONDARY,
  TERTIARY,
  DESTRUCTIVE,
  GHOST
}

enum class ButtonSize {
  SMALL,
  MEDIUM,
  LARGE
}

@Composable
fun AppButton(
  text: String,
  onClick: () -> Unit,
  modifier: Modifier = Modifier,
  variant: ButtonVariant = ButtonVariant.PRIMARY,
  size: ButtonSize = ButtonSize.MEDIUM,
  enabled: Boolean = true,
  loading: Boolean = false,
  icon: @Composable (() -> Unit)? = null,
  fullWidth: Boolean = false,
  contentDescription: String? = null
) {
  val interactionSource = remember { MutableInteractionSource() }
  val isPressed by interactionSource.collectIsPressedAsState()
  
  // Press animation
  val scale by animateFloatAsState(
    targetValue = if (isPressed && enabled) 0.97f else 1f,
    label = "button_press_scale"
  )
  
  // Colors based on variant
  val buttonColors = when (variant) {
    ButtonVariant.PRIMARY -> ButtonDefaults.buttonColors(
      containerColor = MaterialTheme.colorScheme.primary,
      contentColor = MaterialTheme.colorScheme.onPrimary,
      disabledContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.38f),
      disabledContentColor = MaterialTheme.colorScheme.onPrimary.copy(alpha = 0.38f)
    )
    ButtonVariant.SECONDARY -> ButtonDefaults.buttonColors(
      containerColor = MaterialTheme.colorScheme.secondary,
      contentColor = MaterialTheme.colorScheme.onSecondary,
      disabledContainerColor = MaterialTheme.colorScheme.secondary.copy(alpha = 0.38f),
      disabledContentColor = MaterialTheme.colorScheme.onSecondary.copy(alpha = 0.38f)
    )
    ButtonVariant.TERTIARY -> ButtonDefaults.outlinedButtonColors(
      containerColor = Color.Transparent,
      contentColor = MaterialTheme.colorScheme.primary,
      disabledContainerColor = Color.Transparent,
      disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
    )
    ButtonVariant.DESTRUCTIVE -> ButtonDefaults.buttonColors(
      containerColor = MaterialTheme.colorScheme.error,
      contentColor = MaterialTheme.colorScheme.onError,
      disabledContainerColor = MaterialTheme.colorScheme.error.copy(alpha = 0.38f),
      disabledContentColor = MaterialTheme.colorScheme.onError.copy(alpha = 0.38f)
    )
    ButtonVariant.GHOST -> ButtonDefaults.textButtonColors(
      containerColor = Color.Transparent,
      contentColor = MaterialTheme.colorScheme.primary,
      disabledContainerColor = Color.Transparent,
      disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
    )
  }
  
  // Border for tertiary variant
  val border = if (variant == ButtonVariant.TERTIARY) {
    BorderStroke(
      width = BorderWidth.thin,
      color = if (enabled) MaterialTheme.colorScheme.outline else MaterialTheme.colorScheme.outline.copy(alpha = 0.38f)
    )
  } else null
  
  // Height based on size
  val height = when (size) {
    ButtonSize.SMALL -> Sizes.buttonSmall
    ButtonSize.MEDIUM -> Sizes.buttonMedium
    ButtonSize.LARGE -> Sizes.buttonLarge
  }
  
  // Typography based on size
  val textStyle = when (size) {
    ButtonSize.SMALL -> MaterialTheme.typography.labelMedium
    ButtonSize.MEDIUM -> MaterialTheme.typography.labelLarge
    ButtonSize.LARGE -> MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold)
  }
  
  val buttonModifier = modifier
    .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier)
    .height(height)
    .graphicsLayer {
      scaleX = scale
      scaleY = scale
    }
    .semantics {
      role = Role.Button
      contentDescription?.let { this.contentDescription = it }
    }
  
  // Render appropriate button type based on variant
  when (variant) {
    ButtonVariant.TERTIARY -> {
      OutlinedButton(
        onClick = onClick,
        modifier = buttonModifier,
        enabled = enabled && !loading,
        colors = buttonColors,
        border = border,
        interactionSource = interactionSource,
        contentPadding = PaddingValues(horizontal = Spacing.buttonPadding),
        shape = MaterialTheme.shapes.medium
      ) {
        ButtonContent(
          text = text,
          loading = loading,
          icon = icon,
          textStyle = textStyle
        )
      }
    }
    ButtonVariant.GHOST -> {
      TextButton(
        onClick = onClick,
        modifier = buttonModifier,
        enabled = enabled && !loading,
        colors = buttonColors,
        interactionSource = interactionSource,
        contentPadding = PaddingValues(horizontal = Spacing.buttonPadding),
        shape = MaterialTheme.shapes.medium
      ) {
        ButtonContent(
          text = text,
          loading = loading,
          icon = icon,
          textStyle = textStyle
        )
      }
    }
    else -> {
      Button(
        onClick = onClick,
        modifier = buttonModifier,
        enabled = enabled && !loading,
        colors = buttonColors,
        interactionSource = interactionSource,
        contentPadding = PaddingValues(horizontal = Spacing.buttonPadding),
        shape = MaterialTheme.shapes.medium,
        elevation = ButtonDefaults.buttonElevation(
          defaultElevation = Elevation.elevationNone,
          pressedElevation = Elevation.elevationNone
        )
      ) {
        ButtonContent(
          text = text,
          loading = loading,
          icon = icon,
          textStyle = textStyle
        )
      }
    }
  }
}

@Composable
private fun ButtonContent(
  text: String,
  loading: Boolean,
  icon: @Composable (() -> Unit)?,
  textStyle: androidx.compose.ui.text.TextStyle
) {
  if (loading) {
    CircularProgressIndicator(
      modifier = Modifier.size(Sizes.iconSm),
      strokeWidth = 2.dp
    )
  } else {
    Row(
      horizontalArrangement = Arrangement.Center,
      verticalAlignment = Alignment.CenterVertically,
      modifier = Modifier.defaultMinSize(minHeight = Sizes.minTouchTarget)
    ) {
      icon?.let {
        Box(modifier = Modifier.size(Sizes.iconSm)) {
          it()
        }
        Spacer(modifier = Modifier.width(Spacing.spacing8))
      }
      Text(
        text = text,
        style = textStyle
      )
    }
  }
}
